crates/type-system/src/ontology/data_type/mod.rs

```diff
@@ -24,9 +27,11 @@ enum DataTypeTag {
 pub struct DataType {
     #[cfg_attr(target_arch = "wasm32", tsify(type = "'dataType'"))]
     kind: DataTypeTag,
+    #[cfg_attr(target_arch = "wasm32", tsify(type = "`${string}/v/${number}`"))]
```

Comment:
Just want to flag that template literal types would need some sort of validating function to assert structure. Otherwise code using arbitrary non-constant `string`s won't compile. 

I guess we don't currently have a way to write manual typescript methods to add to the package next to the auto generated methods. Is that something we're interested in to have a type assertion validator function for `${string}/v/${number}`?

Detailed explanation:
```typescript
// Imagine this is a user-supplied string we don't have any restriction on
// i.e. it's of type `string`
const maybe$id = "invalidstring/v/";

// We'd need a vaildation functions such as this that asserts the type
const exampleStrIsValidUri = (str: string): str is DataType["$id"] => {
  // This is for brevity. Maybe a regex instead.
  return str.indexOf("/v/") !== -1;
};

// For us to conditionally use it as an $id
if (exampleStrIsValidUri(maybe$id)) {
  const _: DataType = {
    $id: maybe$id,
    // ..
  };
}
```

Reply:
Interesting, lets discuss offline

Reply:
I'm trying out a new approach:
- typing `VersionedUri` by creating a new alias in a private newtype.
- Removing the explicit overrides in the other types, relying on the fact that `Tsify` will keep the original type name if the thing being referred to doesn't implement `Tsify`. So in this instance, **because** `VersionedUri` doesn't implement `Tsify`, any struct with `foo: VersionedUri` will generate a TS interface with `VersionedUri` in the type definition. Which would usually fail TypeScript compilation except we're now adding the newtype under the same name so it resolves nicely.
- Turning `isValidVersionedUri` into a type guard by manually defining the TS function signatures in static strings.
 - This means I was able to get rid of the `allow(non_snake_case)`
   - I'm not sure if we want to repeat this pattern in the other WASM modules, it does mean we have more things to keep in sync, we can't just rely on `wasm_bindgen` creating nice signatures and interpolating the Rust docs in. @TimDiekmann pinging in case this is of interest

https://github.com/blockprotocol/blockprotocol/pull/519/files/0c2ac1574f74c9b09b9898188e210d2399fb0004..becc2fe048b64b0552ef5e4f583e9502551d593b